chemprop/hyperparameter_optimization.py

```diff
@@ -79,25 +78,53 @@ def objective(hyperparams: Dict[str, Union[int, float]]) -> float:
         logger.info(f'num params: {num_params:,}')
         logger.info(f'{mean_score} +/- {std_score} {hyper_args.metric}')
 
-        results.append({
-            'mean_score': mean_score,
-            'std_score': std_score,
-            'hyperparams': hyperparams,
-            'num_params': num_params
-        })
-
         # Deal with nan
         if np.isnan(mean_score):
             if hyper_args.dataset_type == 'classification':
                 mean_score = 0
             else:
                 raise ValueError('Can\'t handle nan score for non-classification dataset.')
 
-        return (1 if hyper_args.minimize_score else -1) * mean_score
+        loss = (1 if hyper_args.minimize_score else -1) * mean_score
 
-    fmin(objective, SPACE, algo=tpe.suggest, max_evals=args.num_iters, rstate=np.random.RandomState(args.seed))
+        return {
+            'loss': loss,
+            'status': 'ok',
+            'mean_score': mean_score,
+            'std_score': std_score,
+            'hyperparams': hyperparams,
+            'num_params': num_params
+        }
+
+    if args.manual_trial_dirs is not None:
+        manual_trials = load_manual_trials(args.manual_trial_dirs, SPACE.keys(), args)
+    else:
+        manual_trials = None
+
+    for i in range(args.num_iters):
+        # run fmin and load trials in single steps to allow for parallel operation
+        trials = load_trials(dir_path=args.hyperopt_checkpoint_dir, previous_trials=manual_trials)
+        if len(trials) >= args.num_iters:
+            break
+
+        hyperopt_seed = get_hyperopt_seed(seed=args.seed, dir_path=args.hyperopt_checkpoint_dir)
+
+        fmin(
+            objective,
+            SPACE,
+            algo=partial(tpe.suggest, n_startup_jobs=args.startup_random_iters),
+            max_evals=len(trials) + 1,
+            trials=trials,
+            rstate=np.random.RandomState(hyperopt_seed),
+        )
+
+        # Create a trials object with only the last instance by merging the last data with an empty trials object
```

Comment:
I guess this is the same question as above, why is it necessary to merge something with an empty trials object? Since this is repeated here, I guess it is not a bug but somehow necessary, and I am curious why :)

Reply:
This is a slightly different version than above. I can't slice a Trials object, only the Trials.trials list of dictionaries that make up its constituent data, and I need to extract the last one to save by itself every iteration of the hyperopt search. I can't figure out a way to take a dict from Trials.trials and make it back into a Trials object directly, I can only figure out how to merge that data dict in with an empty Trials(), which is what I'm doing with merge_trials. It's possible that there's a cleaner way to do this, but it requires a better understanding of hyperopt trials and this is not exactly a use case that's in their documentation.

Reply:
Hmmm. Though I actually haven't tried what would happen if I did `Trials(Trials.trials)`. That might create a new instance of Trials with the data. I will test this.

Reply:
Confirmed this does not work.

Reply:
Haha ok well then I think we should just keep it as it is! Thanks for trying, though!